### PR TITLE
Update and add tests

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Also fixes the lack of support for `sc.Scope` subclasses when using instances as "providers".

### CODE REVIEW

File: tests/complex_workflow_test.py [hunk 1]
```diff
@@ -80,26 +80,13 @@ def iofq(
...
-
-def direct_beam() -> DirectBeam:
-    return DirectBeam(np.array(1 / 1.5))
+params = {
+    Raw[SampleRun]: Raw(RawData(data=np.ones(4), monitor1=1.0, monitor2=2.0)),
+    Raw[BackgroundRun]: Raw(RawData(data=np.ones(4) * 1.5, monitor1=1.0, monitor2=4.0)),
+    DetectorMask: DetectorMask(np.array([1, 1, 0, 1])),
+    SolidAngle: SolidAngle(np.array([1.0, 0.5, 0.25, 0.125])),
+    DirectBeam: DirectBeam(np.array(1 / 1.5)),
+}
```
Comment (YooSunYoung) [inline tests/complex_workflow_test.py:89]: I like this way much better...!

File: tests/pipeline_test.py [hunk 2]
```diff
@@ -58,7 +60,47 @@ def provide_int() -> int:
     assert pipeline.compute((float, str)) == (1.5, "3;1.5")
```
Comment (YooSunYoung) [inline tests/pipeline_test.py:60]: Shouldn't it also have
``assert ncall == 1``
??

Comment (SimonHeybrock) [inline tests/pipeline_test.py:60]:
```suggestion
    assert pipeline.compute((float, str)) == (1.5, "3;1.5")
    assert ncall == 1
```

File: tests/pipeline_test.py [hunk 3]
```diff
@@ -21,19 +21,21 @@ def int_float_to_str(x: int, y: float) -> str:
...
     pipeline = sl.Pipeline([int_to_float])
     with pytest.raises(sl.UnsatisfiedRequirement):
         pipeline.compute(float)
 
 
-def test_intermediate_computed_once() -> None:
+def test_intermediate_used_multiple_times_is_computed_only_once() -> None:
     ncall = 0
 
     def provide_int() -> int:
```
Comment (YooSunYoung) [inline tests/pipeline_test.py:41]: Do you maybe want to use `unittest.mock` for counting calls?

Something like this...? Then we can use the call counting method of `Mock`.
```
from unittest.mock import Mock
MockCounter = NewType('MockCounter', Mock)
def provide_int(counter: MockCounter) -> int:
    counter()
    return 3

def test_whatever() -> None:
    mock_counter = Mock(lambda: None)
    pipeline = sl.Pipeline([int_to_float, provide_int, int_float_to_str], params={MockCounter: mock_counter})
    mock_counter.assert_called_once()
```

Comment (SimonHeybrock) [reply tests/pipeline_test.py:41]: We can consider this if we begin to add more complex tests. For now, the manual way seems sufficient?

**Review by YooSunYoung — APPROVED** [review]

**Pull request merged**